Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,218 @@
+or from multiple subsystems.
+
+The emphasis here is on a common generic interface that keeps subsystem
+customization out of the bus infrastructure.
+
+When Should the Ancillary Bus Not Be Used
+=========================================
+
+If there is a need for a more specifically customized bus that provides
+functionality beyond just the matching up of ancillary_devices and
+ancillary_drivers, and sacrifices common generality to extend the bus
+functionality into subsystem specific functions, then the ancillary bus should
+not be used.  A custom bus solution, or some other method should be considered
+in this use case.
+
+Ancillary Devices
+=================
+
+An ancillary_device is created and registered to represent a part of its parent
+device's functionality. It is given a match_name that is used for driver
+binding and a release callback that is invoked when the device is unregistered.
+
+.. code-block:: c
+
+	struct ancillary_device {
+		struct device dev;
+		const char *match_name;
+		void (*release)(struct ancillary_device *);
+		u32 id;
+	};
+
+The ancillary_device is enumerated when it is attached to the bus. The device
+is assigned a unique ID automatically that will be appended to its name. If
+two ancillary_devices both named "foo" are registered onto the bus, they will
+have the device names, "foo.x" and "foo.y", where x and y are unique integers.
+
+Ancillary Device Memory Model and Lifespan
+------------------------------------------
+
+When a kernel driver registers an ancillary_device on the ancillary bus, we will
+use the nomenclature to refer to this kernel driver as a registering driver.  It
+is the entity that will allocate memory for the ancillary_device and register it
+on the ancillary bus.
+
+A parent object, defined in the shared header file, will contain the
+ancillary_device.  It will also contain a pointer to the shared object(s), which
+will also be defined in the shared header.  Both the parent object and the
+shared object(s) will be allocated by the registering driver.  This layout
+allows the ancillary_driver's registering module to perform a container_of()
+call to go from the pointer to the ancillary_device, that is passed during the
+call to the ancillary_driver's probe function, up to the parent object, and then
+have access to the shared object(s).
+
+The memory for the ancillary_device will be freed only in its release()
+callback flow as defined by its registering driver.
+
+The memory for the shared object(s) must have a lifespan equal to, or greater
+than, the lifespan of the memory for the ancillary_device.  The ancillary_driver
+should only consider that this shared object is valid as long as the
+ancillary_device is still registered on the ancillary bus.  It is up to the
+registering driver to manage (e.g. free or keep available) the memory for the
+shared object beyond the life of the ancillary_device.
+
+The ancillary_device cannot have a lifespan longer than the lifespan of the
```

Comment:
again this is also implicit to any linux bus, nothing specific to ancillary bus.

Reply:
@paravmellanox I agree this section is generic and applies to any bus but it was added based on explicit feedback to spell this out in the documentation. So we should keep it.

Reply:
again, not the normal bus model - we have a software object outside of the bus subsystem responsilbe for the memory of the ancillary_device